src/v2/providers/firestore.ts

```diff
@@ -0,0 +1,453 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import * as firestore from "firebase-admin/firestore";
+import * as logger from "../../logger";
+import { ParamsOf } from "../../common/params";
+import { normalizePath } from "../../common/utilities/path";
+import { PathPattern } from "../../common/utilities/path-pattern";
+import { initV2Endpoint, ManifestEndpoint } from "../../runtime/manifest";
+import { Change, CloudEvent, CloudFunction } from "../core";
+import { EventHandlerOptions, getGlobalOptions, optionsToEndpoint } from "../options";
+import {
+  createBeforeSnapshotFromJson,
+  createBeforeSnapshotFromProtobuf,
+  createSnapshotFromJson,
+  createSnapshotFromProtobuf,
+} from "../../common/providers/firestore";
+
+/** @internal */
+export const writtenEventType = "google.cloud.firestore.document.v1.written";
+
+/** @internal */
+export const createdEventType = "google.cloud.firestore.document.v1.created";
+
+/** @internal */
+export const updatedEventType = "google.cloud.firestore.document.v1.updated";
+
+/** @internal */
+export const deletedEventType = "google.cloud.firestore.document.v1.deleted";
+
+// https://github.com/googleapis/google-cloudevents-nodejs/blob/main/cloud/firestore/v1/DocumentEventData.ts
+/** @internal */
+export interface RawFirestoreDocument {
+  name: string;
+  fields: Record<string, any>;
+  createTime: string;
+  updateTime: string;
+}
+
+/** @internal */
+export interface RawFirestoreData {
+  value?: RawFirestoreDocument;
+  oldValue?: RawFirestoreDocument;
+  updateMask?: { fieldPaths: Array<string> };
+}
+
+/** @internal */
+export interface RawFirestoreEvent extends CloudEvent<Uint8Array | RawFirestoreData> {
+  location: string;
+  project: string;
+  database: string;
+  namespace: string;
+  document: string;
+  datacontenttype: string;
+  dataschema: string;
+  time: string;
+}
+
+/** A Firestore DocumentSnapshot */
+export type DocumentSnapshot = firestore.DocumentSnapshot;
+
+/** A Firestore QueryDocumentSnapshot */
+export type QueryDocumentSnapshot = firestore.QueryDocumentSnapshot;
+
+/** A CloudEvent that contains a DocumentSnapshot or a Change<DocumentSnapshot> */
+export interface FirestoreEvent<T, Params = Record<string, string>> extends CloudEvent<T> {
+  /** The location of the Firestore instance */
+  location: string;
+  /** The project identifier */
+  project: string;
+  /** The Firestore database */
+  database: string;
+  /** The Firestore namespace */
+  namespace: string;
+  /** The document path */
+  document: string;
+  /**
+   * An object containing the values of the path patterns.
+   * Only named capture groups will be populated - {key}, {key=*}, {key=**}
+   */
+  params: Params;
+}
+
+/** DocumentOptions extend EventHandlerOptions with provided document and optional database and namespace.  */
+export interface DocumentOptions<Document extends string = string> extends EventHandlerOptions {
```

Comment:
TS Q: I'm not sure why we need `Document extends string` here. Isn't Document always going to be string anyway?

I see similar pattern in our v1 code (`Path extends string`) and I'm not sure why that's there too 🤔

Reply:
It's used so we can enforce strong typing in our code - https://stackoverflow.com/questions/57342018/extending-a-string-in-typescript

I think Thomas was the one that actually put the Path stuff in during a refactor.